src/application/application.ts

```diff
@@ -1,54 +1,61 @@
 import * as uuidv4 from 'uuid/v4'
-import { handleAPIResponse } from '../util/api';
 import { checkStreamReady, errNoStatus, Stream } from '../util/grpc';
+import { API, ExecutionCreateInputs, ExecutionCreateOutputs, EventStreamInputs, Event, ExecutionStreamInputs, Execution, ExecutionStatus } from '../api';
 
 type Options = {
   client
 }
 
 class Application {
   // api gives access to low level gRPC calls.
-  api
+  private api: API
 
-  constructor(options: Options){
-    this.api = options.client;
+  constructor(api: API) {
+    this.api = api;
   }
 
-  listenEvent(req: ListenEventRequest): Stream<EventData> {
-    return this.api.ListenEvent(req)
+  listenEvent(request: EventStreamInputs): Stream<Event> {
+    return this.api.event.Stream(request)
   }
 
-  listenResult(req: ListenResultRequest): Stream<ResultData> {
-    return this.api.ListenResult(req)
+  listenResult(request: ExecutionStreamInputs): Stream<Execution> {
+    return this.api.execution.Stream(request)
   }
 
-  executeTask(req: ExecuteTaskRequest): Promise<ExecuteTaskReply> {
-    return new Promise<ExecuteTaskReply>((resolve, reject) => {
-      this.api.ExecuteTask(req, handleAPIResponse(resolve, reject))
-    })
+  executeTask(request: ExecutionCreateInputs): ExecutionCreateOutputs {
+    return this.api.execution.Create(request)
   }
 
-  executeTaskAndWaitResult(req: ExecuteTaskRequest): Promise<ResultData> {
-    return new Promise<ResultData>((resolve, reject) => {
+  executeTaskAndWaitResult(request: ExecutionCreateInputs): Promise<Execution> {
+    return new Promise<Execution>((resolve, reject) => {
       const id = uuidv4()
-      const stream = this.listenResult({ serviceID: req.serviceID, tagFilters: [id] })
+      const stream = this.listenResult({
+        filter: {
+          instanceHash: request.instanceHash,
+          statuses: [
+            ExecutionStatus.COMPLETED,
+            ExecutionStatus.FAILED,
+          ],
+          tags: [id]
+        }
+      })
         .on('metadata', (metadata) => {
           const err = checkStreamReady(metadata)
           if (err == errNoStatus) return
           if (err) {
             stream.destroy(err)
             return
           }
-          if (req.executionTags) {
-            req.executionTags.push(id)
+          if (request.tags) {
+            request.tags.push(id)
           } else {
-            req.executionTags = [id]
+            request.tags = [id]
           }
-          this.executeTask(req).catch((err) => stream.destroy(err))
+          this.executeTask(request).catch((err) => stream.destroy(err))
         })
         .on('data', (result) => {
           stream.cancel()
-          resolve(result)
+          result.error ? reject(result.error) : resolve(result)
```

Comment:
```suggestion
          result.error ? reject(new Error(result.error)) : resolve(result)
```
Better to reject an error